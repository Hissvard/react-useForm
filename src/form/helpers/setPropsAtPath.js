export const setPropsAtPath = (fieldData, props, path) => {
  let current = fieldData
  let previousBranch
  let usedProps = false
  const size = path.length
  for (let index = 0; index < size; index++) {
    let branch = path[index]
    if (!isNaN(previousBranch) && branch === 'fields') {
      index++
      branch = path[index]
    }

    if (!usedProps && branch === 'items') {
      usedProps = true
      current = current.props
    }

    if (index === size - 1) {
      current[branch] = usedProps ? props.props : props
    } else {
      previousBranch = branch
      current = current[branch]
    }
  }

  return fieldData
}
