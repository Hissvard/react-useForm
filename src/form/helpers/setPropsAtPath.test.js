import { setPropsAtPath } from './setPropsAtPath'

let fieldData, props
beforeEach(() => {
  fieldData = {
    f0: {
      f1: {
        props: {
          items: [
            {
              f3: {
                items: [
                  { f4: {} },
                  { f4: 'second' },
                ],
              },
            },
          ],
        },
      },
    },
  }
  props = {
    props: {
      prop1: 'prop1',
      meta: 'meta',
    },
  }
})

test('basic path', () => {
  expect(setPropsAtPath(fieldData, props, ['f0'])).toMatchSnapshot()
})

test('nested path', () => {
  expect(setPropsAtPath(fieldData, props, ['f0', 'f1'])).toMatchSnapshot()
})

test('path with items', () => {
  expect(setPropsAtPath(fieldData, props, ['f0', 'f1', 'items', '0', 'fields', 'f3'])).toMatchSnapshot()
})

test('path with items with items', () => {
  expect(setPropsAtPath(fieldData, props, ['f0', 'f1', 'items', '0', 'fields', 'f3', 'items', '1', 'f4'])).toMatchSnapshot()
})
