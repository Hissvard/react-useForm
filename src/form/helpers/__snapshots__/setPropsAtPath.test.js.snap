// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic path 1`] = `
Object {
  "f0": Object {
    "props": Object {
      "meta": "meta",
      "prop1": "prop1",
    },
  },
}
`;

exports[`nested path 1`] = `
Object {
  "f0": Object {
    "f1": Object {
      "props": Object {
        "meta": "meta",
        "prop1": "prop1",
      },
    },
  },
}
`;

exports[`path with items 1`] = `
Object {
  "f0": Object {
    "f1": Object {
      "props": Object {
        "items": Array [
          Object {
            "f3": Object {
              "meta": "meta",
              "prop1": "prop1",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`path with items with items 1`] = `
Object {
  "f0": Object {
    "f1": Object {
      "props": Object {
        "items": Array [
          Object {
            "f3": Object {
              "items": Array [
                Object {
                  "f4": Object {},
                },
                Object {
                  "f4": Object {
                    "meta": "meta",
                    "prop1": "prop1",
                  },
                },
              ],
            },
          },
        ],
      },
    },
  },
}
`;
